k2/csrc/fsa.h

```diff
@@ -77,13 +77,40 @@ struct Fsa {
   // contains the first arc-index leaving this state (index into `arcs`).
   // The next element of this array gives the end of that range.
   // Note: the final-state is numbered last, and implicitly has no
-  // arcs leaving it.
+  // arcs leaving it. For non-empty FSA, we put a duplicate of the final state
+  // at the end of `arc_indexes` to avoid boundary check for some FSA
+  // operations. Caution: users should never call `arc_indexes.size()` to get
+  // the number of states, they should call `NumStates()` to get the number.
   std::vector<int32_t> arc_indexes;
 
   // Note: an index into the `arcs` array is called an arc-index.
   std::vector<Arc> arcs;
 
-  StateId NumStates() const { return static_cast<StateId>(arc_indexes.size()); }
+  Fsa() {}
+  // just for creating testing FSA examples for now.
+  Fsa(std::vector<Arc> fsa_arcs, int32_t final_state) {
```

Comment:
The constructor should handle the case that the arcs are not sorted by the `src_state`.

Reply:
Actually I don't think so as:
- our `arcs` are always sorted state by state, many implentation are based on this assumption. This constructor is not reading arcs from file or strings, it's only vector. So we can suppose that it is always sorted as this vector is either created by users manually or created by other algorithms (also sorted).
- for now this constructor is only for testing, it's no necessary to write non-sorted cases which will make hard to read.

Reply:
Maybe @danpovey have something to say about this.